src/human_encoding/parse/ast.rs

```diff
@@ -0,0 +1,846 @@
+    fn from_2<T, U, F1, F2, F>(toks: &mut [Self], convert1: F1, convert2: F2, unconvert: F) -> Self
+    where
+        F1: FnOnce(&mut Self) -> T,
+        F2: FnOnce(&mut Self) -> U,
+        F: FnOnce(T, U) -> Self,
+    {
+        if let Some(e) = propagate_errors(toks) {
+            return e;
+        }
+        assert_eq!(toks.len(), 3);
+        toks[1].expect_position();
+        unconvert(convert1(&mut toks[0]), convert2(&mut toks[2]))
+    }
+
+    /// Creates an `Ast` from three sub-`Ast`s with dummies in between
+    fn from_3<T, U, V, F1, F2, F3, F>(
+        toks: &mut [Self],
+        convert1: F1,
+        convert2: F2,
+        convert3: F3,
+        unconvert: F,
+    ) -> Self
+    where
+        F1: FnOnce(&mut Self) -> T,
+        F2: FnOnce(&mut Self) -> U,
+        F3: FnOnce(&mut Self) -> V,
+        F: FnOnce(T, U, V) -> Self,
+    {
+        if let Some(e) = propagate_errors(toks) {
+            return e;
+        }
+        assert_eq!(toks.len(), 5);
+        toks[1].expect_position();
+        toks[3].expect_position();
+        unconvert(
+            convert1(&mut toks[0]),
+            convert2(&mut toks[2]),
+            convert3(&mut toks[4]),
+        )
+    }
+
+    /// Creates an AST from a combinator with no arguments
+    fn from_combinator(mut toks: Vec<Self>) -> Self {
+        if let Some(e) = propagate_errors(&toks) {
+            return e;
+        }
+
+        let (comb, position) = match mem::replace(&mut toks[0], Ast::Replaced) {
+            Ast::Combinator { comb, position } => (comb, position),
+            _ => unreachable!(),
+        };
+
+        // This stupid construction is needed to avoid borrowck rules
+        // around borrowing tok[1] and tok[2] mutably at the same time.
+        let arcs: Vec<Arc<_>> = toks[1..]
+            .iter_mut()
+            .map(Ast::expect_expression)
+            .map(Arc::new)
+            .collect();
+        let inner = comb.map_left_right(|_| Arc::clone(&arcs[0]), |_| Arc::clone(&arcs[1]));
+        Ast::Expression(Expression {
+            inner: ExprInner::Inline(inner),
+            position,
+        })
+    }
+
+    /// Creates an AST from a dummy lexeme
+    fn from_dummy_lexeme(lexemes: &[&std::rc::Rc<Lexeme>]) -> Self {
+        assert_eq!(lexemes.len(), 1);
+        let position = (&lexemes[0].position).into();
+        Ast::Dummy { position }
+    }
+
+    /// Creates an AST from a lexeme which forms a complete type
+    fn from_type_lexeme(lexemes: &[&std::rc::Rc<Lexeme>]) -> Self {
+        assert_eq!(lexemes.len(), 1);
+        let position = &lexemes[0].position;
+        match lexemes[0].raw.as_str() {
+            "1" => Ast::Type(Some(Type::One)),
+            "2" => Ast::Type(Some(Type::Two)),
+            other => {
+                assert_eq!(&other[..2], "2^");
+                match str::parse::<u32>(&other[2..]) {
+                    // TODO how many of these should we support?
```

Comment:
8e711fd6b8a34c454d7ec83a8ad672cc6a8e12f1: Ideally we should support as many powers as in `types::precomputed::N_POWERS`, but the numbers get pretty big. We could add SI prefixes for kibi and mebi. Maybe in another PR.

Reply:
Will leave for a later PR.